server/service/src/sync/translations/special/name_merge.rs

```diff
@@ -56,7 +60,57 @@ impl SyncTranslation for NameMergeTranslation {
             })
             .collect();
 
-        Ok(Some(IntegrationRecords::from_upserts(upsert_records)))
+        let name_store_join_repo = NameStoreJoinRepository::new(connection);
+        let name_store_joins_for_delete = name_store_join_repo.query(Some(
+            NameStoreJoinFilter::new().name_id(EqualFilter::equal_to(&data.merge_id_to_delete)),
+        ))?;
+        let name_store_joins_for_keep = name_store_join_repo.query(Some(
+            NameStoreJoinFilter::new().name_id(EqualFilter::equal_to(&data.merge_id_to_keep)),
+        ))?;
+        let mut deletes: Vec<PullDeleteRecord> = vec![];
+
+        // We need to delete the name_store_joins that are no longer needed after the merge
+        // Situation A: ("Joined to" meaning store->nsj->name_link->name)
+        // storeA joined to nameK
+        // storeA joined to nameD
+        // storeB joined to nameD
+        // nameD merged into nameK
+        // storeA joined to nameK
+        // storeA joined to nameK (delete this join to avoid showing twice in lists seemingly as a duplicate)
+        // storeB joined to nameK (make sure we don't accidentally delete this one, or visibility of nameK will be lost for storeB)
+        name_store_joins_for_delete.iter().for_each(|nsj_delete| {
+            if name_store_joins_for_keep.iter().any(|nsj_keep| {
+                nsj_keep.name_store_join.store_id == nsj_delete.name_store_join.store_id
+            }) {
+                deletes.push(PullDeleteRecord {
+                    id: nsj_delete.name_store_join.id.clone(),
+                    table: PullDeleteRecordTable::NameStoreJoin,
+                });
+            }
+        });
+
+        // Situation B:
+        // storeK.name_id == nameK.id
+        // storeK joined to nameD
+        // nameD merged into nameK
+        // storeK joined to nameK (delete the join before this happens, stores shouldn't be visible to themselves)
+        let store_repo = StoreRepository::new(connection);
+        let store_filter =
+            StoreFilter::new().name_id(EqualFilter::equal_to(&data.merge_id_to_keep));
+        let store = store_repo.query_one(store_filter)?;
+
+        if let Some(store) = store {
+            name_store_joins_for_delete.iter().for_each(|nsj_delete| {
+                if nsj_delete.name_store_join.store_id == store.store_row.id {
+                    deletes.push(PullDeleteRecord {
+                        id: nsj_delete.name_store_join.id.clone(),
+                        table: PullDeleteRecordTable::NameStoreJoin,
+                    });
+                }
+            });
+        }
```

Comment:
Found it hard to understand when you have two loops. For example, would it be possible to add the same nsj_delete to the `delete` Vec twice? (no but its hard to see)


To avoid it you could do it in a single filter_map():
```rust
let deletes = name_store_joins_for_delete
            .iter()
            .filter_map(|nsj_delete| {
                // delete nsj_delete if it points to the store that belongs to the "keep" name
                if let Some(store) = &store {
                    if nsj_delete.name_store_join.store_id == store.store_row.id {
                        return Some(PullDeleteRecord {
                            id: nsj_delete.name_store_join.id.clone(),
                            table: PullDeleteRecordTable::NameStoreJoin,
                        });
                    }
                }

                // delete dublicates
                if name_store_joins_for_keep.iter().any(|nsj_keep| {
                    nsj_keep.name_store_join.store_id == nsj_delete.name_store_join.store_id
                }) {
                    return Some(PullDeleteRecord {
                        id: nsj_delete.name_store_join.id.clone(),
                        table: PullDeleteRecordTable::NameStoreJoin,
                    });
                }

                None
            })
            .collect::<Vec<_>>();
```

Reply:
Heh I actually tried reading how to use filter_map but found it confusing to get the syntax right as I've never used it before. So perhaps more readable if familiar with the esoteric rust dozens of [built in methods](https://doc.rust-lang.org/std/iter/trait.Iterator.html#provided-methods) that exist on `Iter` alone? 😛 

I actually like it though looks neat! Thanks. Just whinging about there always being 10 different ways to shave a yak in rust. Golang gang 🔫 🔫 🔫 🔫 🔫

Reply:
Hmmm this might get gross now I've added an upsert too lol

Reply:
yeah agree too many ways to do things. These methods also don't work well with async and handling Result is also a bit "magic"...

You can still merge it in the for_each() though, i.e. only iterate the `name_store_joins_for_delete` once